web_ui/frontend/helpers/login.tsx

```diff
@@ -1,8 +1,81 @@
-export async function isLoggedIn() {
-    let response = await fetch("/api/v1.0/auth/whoami")
-    if(!response.ok){
+
+export interface Authenticated {
+    authenticated: boolean
+    csrf_token: string
+    role: string
+    time: number
+    user: string
+}
+
+export async function secureFetch(url: string | URL, options: RequestInit = {}) {
+    if(await isLoggedIn()) {
+
+        // If they are logged in, this key must exist
+        const authenticated = getJsonFromSessionStorage<Authenticated>("authenticated") as Authenticated
+
+        return await fetch(url, {
+            ...options,
+            headers: {
+                ...options.headers,
+                "X-CSRF-Token": authenticated.csrf_token
+            }
+        })
+    }
+
+    throw new Error("You must be logged in to make this request")
+}
+
+export function getJsonFromSessionStorage<O>(key: string) : O | null {
+    if(sessionStorage.getItem(key) !== null) {
+        return JSON.parse(sessionStorage.getItem(key) as string)
+    }
+    return null
+}
+
+export function getAuthenticated() : Authenticated | null {
+    return getJsonFromSessionStorage<Authenticated>("authenticated")
+}
+
+// Allow them to see a page if logged in
+export async function isLoggedIn() : Promise<boolean> {
+
+    // If the session is valid then read it
+    const authenticated = getJsonFromSessionStorage<Authenticated>("authenticated")
+    if(authenticated != null){
+        if(authenticated.time + 10000 > Date.now()){
+            return authenticated.authenticated
+        }
+    }
+
+    // Check if the user is authenticated
+    try {
+
+        let response = await fetch("/api/v1.0/auth/whoami")
+        if(!response.ok){
+            return false
+        }
+
+        const json = await response.json()
+        const authenticated = json['authenticated']
+
+        // If authenticated, store status and csrf token
+        if(authenticated){
+            sessionStorage.setItem(
+                "authenticated",
+                JSON.stringify({
+                    time: Date.now(),
+                    authenticated: true,
+                    user: json['user'],
+                    role: json['role'],
+                    csrf_token: response.headers.get('X-CSRF-Token')
```

Comment:
Yeah, this token isn't really secure in any sense. The only way it could be abuse if we had a XSS vulnerability in which case they could do it even without it in sessionStorage.

Reply:
The only point of the token is to confirm that the request is sent from our site. If they can steal it from sessionStorage they can send requests from our site as well so they could just get their own token and use it.